presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/raw/RawRowEncoder.java

```diff
@@ -0,0 +1,282 @@
+
+            valueLengths[i] = parseMapping(columnHandle, fieldType);
+            capacity += valueLengths[i];
+
+            checkArgument(isSupportedType(columnHandle.getType()), "Unsupported column type '%s' for column '%s'", columnHandle.getType().getDisplayName(), columnHandle.getName());
+        }
+
+        this.buffer = ByteBuffer.allocate(capacity);
+        this.currentBufferPosition = 0;
+    }
+
+    private FieldType parseFieldType(EncoderColumnHandle columnHandle)
+    {
+        try {
+            return Optional.ofNullable(columnHandle.getDataFormat())
+                    .map(dataFormat -> FieldType.valueOf(dataFormat.toUpperCase(Locale.ENGLISH)))
+                    .orElse(FieldType.BYTE);
+        }
+        catch (IllegalArgumentException e) {
+            throw new IllegalArgumentException(format("Invalid dataFormat '%s' for column '%s'", columnHandle.getDataFormat(), columnHandle.getName()));
+        }
+    }
+
+    private void checkFieldType(EncoderColumnHandle columnHandle, FieldType fieldType)
+    {
+        String columnName = columnHandle.getName();
+        Type columnType = columnHandle.getType();
+        if (columnType == BIGINT) {
+            checkFieldTypeOneOf(fieldType, columnName, columnType, FieldType.BYTE, FieldType.SHORT, FieldType.INT, FieldType.LONG);
+        }
+        else if (columnType == INTEGER) {
+            checkFieldTypeOneOf(fieldType, columnName, columnType, FieldType.BYTE, FieldType.SHORT, FieldType.INT);
+        }
+        else if (columnType == SMALLINT) {
+            checkFieldTypeOneOf(fieldType, columnName, columnType, FieldType.BYTE, FieldType.SHORT);
+        }
+        else if (columnType == TINYINT) {
+            checkFieldTypeOneOf(fieldType, columnName, columnType, FieldType.BYTE);
+        }
+        else if (columnType == BOOLEAN) {
+            checkFieldTypeOneOf(fieldType, columnName, columnType, FieldType.BYTE, FieldType.SHORT, FieldType.INT, FieldType.LONG);
+        }
+        else if (columnType == DOUBLE) {
+            checkFieldTypeOneOf(fieldType, columnName, columnType, FieldType.DOUBLE, FieldType.FLOAT);
+        }
+        else if (isVarcharType(columnType)) {
+            checkFieldTypeOneOf(fieldType, columnName, columnType, FieldType.BYTE);
+        }
+    }
+
+    private void checkFieldTypeOneOf(FieldType declaredFieldType, String columnName, Type columnType, FieldType... allowedFieldTypes)
+    {
+        if (!Arrays.asList(allowedFieldTypes).contains(declaredFieldType)) {
+            throw new IllegalArgumentException(format(
+                    "Wrong dataFormat '%s' specified for column '%s'; %s type implies use of %s",
+                    declaredFieldType.name(),
+                    columnName,
+                    columnType.getDisplayName(),
+                    Joiner.on("/").join(allowedFieldTypes)));
+        }
+    }
+
+    private int parseMapping(EncoderColumnHandle columnHandle, FieldType fieldType)
```

Comment:
can we make it `@VisibleForTesting` and add unit test for it?

Reply:
I could test it without `@VisibleForTesting` annotation just by calling encoder factory create method on columns with different mappings

Reply:
but that wouldn't actually test the method

Reply:
Given logic complexity I would probably opt toward directly unit testing method. But if you can make it look nice and readable testing at wider scope. And still be able to excercise all corner cases, it should be fine too.